crucible-mir/src/Mir/TransCustom.hs

```diff
@@ -1132,6 +1134,68 @@ size_of = (["core", "intrinsics", "size_of"], \substs -> case substs of
     _ -> Nothing
     )
 
+size_of_val :: (ExplodedDefId, CustomRHS)
+size_of_val = (["core", "intrinsics", "size_of_val"], \substs -> case substs of
+    Substs [_] -> Just $ CustomOp $ \opTys ops -> case (opTys, ops) of
+        -- We first check whether the underlying type is sized or not.
+        ([TyRawPtr ty Immut], [MirExp tpr e]) -> case tpr of
+            -- Slices (e.g., `&[u8]`, `&str`, and custom DSTs whose last field
+            -- contains a slice) are unsized. We currently support computing
+            -- the size of slice values that aren't embedded in a custom DST.
+            MirSliceRepr -> case ty of
+                TySlice elemTy -> sizeOfSlice elemTy e
+                TyStr {} -> sizeOfSlice (TyUint B8) e
+                TyAdt {} -> unsupportedCustomDst ty
+                _ -> panic "size_of_val"
+                       ["Unexpected MirSliceRepr type", show (PP.pretty ty)]
+
+            -- Trait objects (e.g., `&dyn Debug` and custom DSTs whose last
+            -- field contains a trait object) are unsized. This override
+            -- currently does not support any kind of trait object, so all we
+            -- do here is make an effort to give a descriptive error message.
+            DynRefRepr -> case ty of
+                TyDynamic {} -> unsupportedTraitObject ty
+                TyAdt {} -> unsupportedCustomDst ty
+                _ -> panic "size_of_val"
+                       ["Unexpected DynRefRepr type", show (PP.pretty ty)]
+
+            -- All other cases should correspond to sized types. For these
+            -- cases, computing the value's size is equivalent to computing the
+            -- type's size.
+            MirReferenceRepr ->
+                getLayoutFieldAsMirExp "size_of_val" laySize ty
```

Comment:
Extern types (`TyForeign`), which are unsized but have no metadata, would presumably also fall into this case.  Extern types are unstable, and it sounds like the Rust devs are likely to implement the "`Sized` hierarchy" feature first, which will (among other things) statically prevent calling `size_of_val` on extern types.  So we can probably get away with ignoring them here.